src/traces/scatter/plot.js

```diff
@@ -157,12 +153,17 @@ function plotOne(gd, idx, plotinfo, cdscatter, cdscatterAll, element, transition
     var xa = plotinfo.xaxis,
         ya = plotinfo.yaxis;
 
-    var trace = cdscatter[0].trace,
-        line = trace.line,
-        tr = d3.select(element);
+    var trace = cdscatter[0].trace;
+    var line = trace.line;
+    var tr = d3.select(element);
+
+    var errorBarGroup = ensureSingle(tr, 'g', 'errorbars');
+    var lines = ensureSingle(tr, 'g', 'lines');
+    var points = ensureSingle(tr, 'g', 'points');
+    var text = ensureSingle(tr, 'g', 'text');
```

Comment:
Hmm. So now all scatter traces (no-matter their `mode`, `fill` or `error_x`/`error_y` settings) get four sub `<g>`:

![image](https://user-images.githubusercontent.com/6675409/45224769-ce39f000-b288-11e8-9ce1-3648051f9444.png)

What if instead we made a "mode" groups data-bind:

```js
var modeData = []
if(trace.error_y.visible || trace.error_x.visible) modeData.push('errorbars');
if(trace.mode.indexOf('lines') !== -1) modeData.push('lines');
if(trace.mode.indexOf('markers') !== -1) modeData.push('points');
if(trace.mode.indexOf('text') !== -1) modeData.push('text');

 var modeGroups = tr.selectAll('g.mode'))
        .data(modeData, identity);
 modeGroups.exit().remove();
 modeGroups.enter().append('g')
        .classed('mode', true)
        .attr('class', identity);
 modeGroups.order();
```

That way those empty `<g>` would be gone :small_airplane:

Reply:
(discussed offline but for future reference)

I tried to do this... and it was going well for a while, but I ran into issues with the fact that various disjoint pathways (eg styling, errorbars) expect the trace data to be attached to these groups. And if you use `selection.data` to control creation of these groups, you can't use it to also pass data down to its children. We could nest another single group inside, just for the purpose of passing down the data, but that seems like it's creating exactly the same kind of problem we're trying to avoid. It's almost like we want a parallel data binding, like extending d3 to allow selecting & operating based on a different attribute from `__data__`... or perhaps we make a separate `Lib` function specifically for this purpose: ensure the ordering and conditional existence of a bunch of groups, but bind the same data to each group.

Anyway, for now we agreed to leave it as is.

Reply:
> Anyway, for now we agreed to leave it as is.

:ok_hand: